velox/common/base/CompareFlags.h

```diff
@@ -14,38 +14,137 @@
-  }
+     Under this mode, result of comparison can be indeterminate; indeterminate
+     result is represented as std::nullopt, and means that the function can not
+     decide on the result of the comparison due to some existing nulls. But not
+     every null results in indeterminate result.
+
+     ## When equalsOnly=true:
+     The compare can return kIndeterminate, or a value according to the
+     following:
+        1. Primitive types and top level nulls:
+          - Comparing null with anything else is indeterminate, otherwise a
+            value result is returned.
+
+          - Comapring top level nulls in complex types is indeterminate.
+
+        2. Arrays:
+          - If the compared array sizes are different then result is false,
+            ex:[null] == [null, null] is false.
+
+           - If any two elements compare result is false, result is false
+             ex: [null, 1] = [null, 2] is false.
+
+           - If result is not false and any two elements compared result is
+             indeterminate, then result is indeterminate.
+             ex: [null, 1] = [null, 1] is indeterminate.
+
+           - If if elements comapre results are true, then result is true.
+             ex: [1, 1] = [1, 1] is true.
+
+        3. Rows: Follows the same logic as array, with fields being elements.
+          - If any two fields compare result is false, result is false
+            ex: (null, 1) = (null, 2) is false.
+
+          - If result is not false and any two fields compare result is
+            indeterminate, then the result is indeterminate.
+            ex: (null, 1) = (null, 1) is indeterminate.
+
+            - If all fields comapre results are true, then result is true.
+            ex: (1, 1) = (1, 1) is indeterminate.
+
+        4. Maps:
+          - Keys are compared first, if keys are not equal values are not
+            checked even if they have nulls and result is false.
+
+         - If keys are the same for all maps, values are compared by applying
+           the array compare logic explained above observing each map as an
+           array of values.
+           ex: {1:null, 2:2} = {1:null, 2:3} is false.
+           ex: {1:null, 2:2} = {1:null, 2:2} is indeterminate.
+           ex: {1:1, 2:2} = {1:1, 2:2} is true.
+
+     ## When equalsOnly=false:
+      The compare either returns a value or throws a user error if a null is
+      encountered before result is determined, as explained below:
+
+        1. Primitive types and top level nulls:
+          - Comparing null with anything else throws (note that functions like
+            >, < do not pass nulls to compare since they are default nulls
+            functions).
+
+          - Comparing top level nulls also throws.
+
+        2. Arrays:
+          - Only elements up to index min(rhs.size(), lhs.size()) are compared.
+
+          - Elements are compared in order starting from index 0.
```

Comment:
It might be simpler combine the first 2 bullets and say " Elements are compared in order starting from index 0 to min(rhs.size(), lhs.size())".

Its only after I read all the remaining bullet points could I understand why the first one was relevant.